pkg/build/strategies/sti/postexecutorstep.go

```diff
@@ -482,3 +458,101 @@ func createCommandForExecutingRunScript(scriptsURL map[string]string, location s
 	}
 	return cmd
 }
+
+func downloadAndExtractFileFromContainer(docker dockerpkg.Docker, tar s2itar.Tar, sourcePath, destinationPath, containerID string) (api.FailureReason, error) {
+	glog.V(5).Infof("Downloading file %q", sourcePath)
+
+	fd, err := ioutil.TempFile(destinationPath, "s2i-runtime-artifact")
+	if err != nil {
+		res := utilstatus.NewFailureReason(
+			utilstatus.ReasonFSOperationFailed,
+			utilstatus.ReasonMessageFSOperationFailed,
+		)
+		return res, fmt.Errorf("could not create temporary file for runtime artifact: %v", err)
+	}
+	defer func() {
+		fd.Close()
+		os.Remove(fd.Name())
+	}()
+
+	if err := docker.DownloadFromContainer(sourcePath, fd, containerID); err != nil {
+		res := utilstatus.NewFailureReason(
+			utilstatus.ReasonGenericS2IBuildFailed,
+			utilstatus.ReasonMessageGenericS2iBuildFailed,
+		)
+		return res, fmt.Errorf("could not download file (%q -> %q) from container %s: %v", sourcePath, fd.Name(), containerID, err)
+	}
+
+	// after writing to the file descriptor we need to rewind pointer to the beginning of the file before next reading
+	if _, err := fd.Seek(0, os.SEEK_SET); err != nil {
+		res := utilstatus.NewFailureReason(
+			utilstatus.ReasonGenericS2IBuildFailed,
+			utilstatus.ReasonMessageGenericS2iBuildFailed,
+		)
+		return res, fmt.Errorf("could not seek to the beginning of the file %q: %v", fd.Name(), err)
+	}
+
+	if err := tar.ExtractTarStream(destinationPath, fd); err != nil {
+		res := utilstatus.NewFailureReason(
+			utilstatus.ReasonGenericS2IBuildFailed,
+			utilstatus.ReasonMessageGenericS2iBuildFailed,
+		)
+		return res, fmt.Errorf("could not extract runtime artifact %q into the directory %q: %v", sourcePath, destinationPath, err)
+	}
+
+	return utilstatus.NewFailureReason("", ""), nil
+}
+
+// check for new labels and apply to the output image.
+func checkAndGetNewLabels(builder *STI, docker dockerpkg.Docker, tar s2itar.Tar, containerID string) error {
+	glog.V(3).Infof("Checking for new Labels to apply... ")
+
+	// metadata filename and its path inside the container
+	metadataFilename := "image_metadata.json"
+	sourceFilepath := filepath.Join("/opt/docker", metadataFilename)
+
+	// create the 'downloadPath' folder if it doesn't exist
+	downloadPath := filepath.Join(builder.config.WorkingDir, "metadata")
+	glog.V(3).Infof("Creating the download path '%s'", downloadPath)
+	if err := os.MkdirAll(downloadPath, 0700); err != nil {
+		glog.Errorf("Error creating dir %q for '%s': %v", downloadPath, metadataFilename, err)
+		return err
+	}
+
+	// download & extract the file from container
+	_, err := downloadAndExtractFileFromContainer(docker, tar, sourceFilepath, downloadPath, containerID)
```

Comment:
I think that you could go ahead and wrap this in the if _, err := ......); err != nil and then return nil inside of the if, then get rid of the else { and return nil at the end.